File: lab3/Stack0.5/stack_adt/tb.pl
#================================================================--
# Design Unit  : cew testbench for stack
#
# File Name    : tb.cew
#
# Purpose      : unit testing
#
# Note         :
#
# Limitations  :
#
# Errors       : none known
#
# Modules      : stack_adt::stack
#
# Dependences  : cew
#
# Author       : Peter Walsh, Vancouver Island University
#
# System       : Perl (Linux)
#
#------------------------------------------------------------------
# Revision List
# Version      Author  Date    Changes
# 1.0          PW      Oct 12  New version
#================================================================--

$|=1;
use strict;
use warnings;











use lib '../';
use stack_adt::stack;
use exc::exception;
use Try::Tiny;

my $cew_Test_Count=0;
          my $cew_Error_Count=0;


# Local Function Load (s, n);
# pushes n values on the stack from
# the sequence 10, 100, 1000 ... 
# note: no exception checking

sub load {
   my $s=shift @_;
   my $n=shift @_;
   
   for (my $i=0; $i<$n; $i++) {
      $s->push(($i+1)*10);
   }
}

#############
# empty stack
#############

my $stack0=stack_adt::stack->new(10);

          do {
              $cew_Test_Count=$cew_Test_Count+1;
              try {
                  $stack0->top();  
                  $cew_Error_Count=$cew_Error_Count+1;
                  print("Test Case ERROR (Ecase) in script at line number ", 60, "\n");
                  print ( "Expected exception full not thrown\n");
              }
              catch {
                  my $error = $_;
                  if (ref($error) eq "exc::exception") {
                      if ($error->get_exc_name() eq "empty") {
                      }
                      else {
                          $cew_Error_Count=$cew_Error_Count+1;  
                          print("Test Case ERROR (Ecase) in script at line number ", 60, "\n");
                          print ( "Unexpected exception ", "empty", " thrown\n");                  
                      }
                  }
               }
           };


          do {
              $cew_Test_Count=$cew_Test_Count+1;
              try {
                  $stack0->pop();  
                  $cew_Error_Count=$cew_Error_Count+1;
                  print("Test Case ERROR (Ecase) in script at line number ", 61, "\n");
                  print ( "Expected exception full not thrown\n");
              }
              catch {
                  my $error = $_;
                  if (ref($error) eq "exc::exception") {
                      if ($error->get_exc_name() eq "empty") {
                      }
                      else {
                          $cew_Error_Count=$cew_Error_Count+1;  
                          print("Test Case ERROR (Ecase) in script at line number ", 61, "\n");
                          print ( "Unexpected exception ", "empty", " thrown\n");                  
                      }
                  }
               }
           };


#################
# half full stack
#################

my $stack1=stack_adt::stack->new(10);

          do {
            $cew_Test_Count=$cew_Test_Count+1;
            try {
                load($stack1, 5) ;
                if (($stack1->top()) != (50)) {
                    $cew_Error_Count=$cew_Error_Count+1;
                    print("Test Case ERROR (Ncase) in script at line number ", 68, "\n");
                    print("Actual Value is ", $stack1->top(), " \n");
                    print("Expected Value is ", 50, "\n");
                }
             }
             catch{
                $cew_Error_Count=$cew_Error_Count+1;
                my $except = $_;
                if (ref($except) eq "exc::exception") {
                    print("Test Case ERROR (Ncase) in script at line number ", 68, "\n");
                    print("Unexpected Exception full thrown\n");
                }
             }
          };


          do {
            $cew_Test_Count=$cew_Test_Count+1;
            try {
                $stack1->pop() ;
                if (($stack1->top()) != (40)) {
                    $cew_Error_Count=$cew_Error_Count+1;
                    print("Test Case ERROR (Ncase) in script at line number ", 69, "\n");
                    print("Actual Value is ", $stack1->top(), " \n");
                    print("Expected Value is ", 40, "\n");
                }
             }
             catch{
                $cew_Error_Count=$cew_Error_Count+1;
                my $except = $_;
                if (ref($except) eq "exc::exception") {
                    print("Test Case ERROR (Ncase) in script at line number ", 69, "\n");
                    print("Unexpected Exception full thrown\n");
                }
             }
          };


          do {
            $cew_Test_Count=$cew_Test_Count+1;
            try {
                $stack1->pop() ;
                if (($stack1->top()) != (30)) {
                    $cew_Error_Count=$cew_Error_Count+1;
                    print("Test Case ERROR (Ncase) in script at line number ", 70, "\n");
                    print("Actual Value is ", $stack1->top(), " \n");
                    print("Expected Value is ", 30, "\n");
                }
             }
             catch{
                $cew_Error_Count=$cew_Error_Count+1;
                my $except = $_;
                if (ref($except) eq "exc::exception") {
                    print("Test Case ERROR (Ncase) in script at line number ", 70, "\n");
                    print("Unexpected Exception full thrown\n");
                }
             }
          };


#################
# full stack
#################

my $stack2=stack_adt::stack->new(10);

          do {
            $cew_Test_Count=$cew_Test_Count+1;
            try {
                load($stack2, 10) ;
                if (($stack2->top()) != (100)) {
                    $cew_Error_Count=$cew_Error_Count+1;
                    print("Test Case ERROR (Ncase) in script at line number ", 77, "\n");
                    print("Actual Value is ", $stack2->top(), " \n");
                    print("Expected Value is ", 100, "\n");
                }
             }
             catch{
                $cew_Error_Count=$cew_Error_Count+1;
                my $except = $_;
                if (ref($except) eq "exc::exception") {
                    print("Test Case ERROR (Ncase) in script at line number ", 77, "\n");
                    print("Unexpected Exception full thrown\n");
                }
             }
          };


          do {
              $cew_Test_Count=$cew_Test_Count+1;
              try {
                  ;  
                  $cew_Error_Count=$cew_Error_Count+1;
                  print("Test Case ERROR (Ecase) in script at line number ", 78, "\n");
                  print ( "Expected exception full not thrown\n");
              }
              catch {
                  my $error = $_;
                  if (ref($error) eq "exc::exception") {
                      if ($error->get_exc_name() eq "full") {
                      }
                      else {
                          $cew_Error_Count=$cew_Error_Count+1;  
                          print("Test Case ERROR (Ecase) in script at line number ", 78, "\n");
                          print ( "Unexpected exception ", "full", " thrown\n");                  
                      }
                  }
               }
           };


          do {
              $cew_Test_Count=$cew_Test_Count+1;
              try {
                  $stack2->push(110);  
                  $cew_Error_Count=$cew_Error_Count+1;
                  print("Test Case ERROR (Ecase) in script at line number ", 79, "\n");
                  print ( "Expected exception full not thrown\n");
              }
              catch {
                  my $error = $_;
                  if (ref($error) eq "exc::exception") {
                      if ($error->get_exc_name() eq "full") {
                      }
                      else {
                          $cew_Error_Count=$cew_Error_Count+1;  
                          print("Test Case ERROR (Ecase) in script at line number ", 79, "\n");
                          print ( "Unexpected exception ", "full", " thrown\n");                  
                      }
                  }
               }
           };


          do {
            $cew_Test_Count=$cew_Test_Count+1;
            try {
                 ;
                if (($stack2->top()) != (100)) {
                    $cew_Error_Count=$cew_Error_Count+1;
                    print("Test Case ERROR (Ncase) in script at line number ", 80, "\n");
                    print("Actual Value is ", $stack2->top(), " \n");
                    print("Expected Value is ", 100, "\n");
                }
             }
             catch{
                $cew_Error_Count=$cew_Error_Count+1;
                my $except = $_;
                if (ref($except) eq "exc::exception") {
                    print("Test Case ERROR (Ncase) in script at line number ", 80, "\n");
                    print("Unexpected Exception full thrown\n");
                }
             }
          };


          do {
            $cew_Test_Count=$cew_Test_Count+1;
            try {
                $stack2->pop() ;
                if (($stack2->top()) != (90)) {
                    $cew_Error_Count=$cew_Error_Count+1;
                    print("Test Case ERROR (Ncase) in script at line number ", 81, "\n");
                    print("Actual Value is ", $stack2->top(), " \n");
                    print("Expected Value is ", 90, "\n");
                }
             }
             catch{
                $cew_Error_Count=$cew_Error_Count+1;
                my $except = $_;
                if (ref($except) eq "exc::exception") {
                    print("Test Case ERROR (Ncase) in script at line number ", 81, "\n");
                    print("Unexpected Exception full thrown\n");
                }
             }
          };


################
# stress test
################

my $stack3=stack_adt::stack->new(100);

for (my $i=0; $i<100; $i++) {
   
          do {
            $cew_Test_Count=$cew_Test_Count+1;
            try {
                $stack3->push($i) ;
                if (($stack3->top()) != ($i)) {
                    $cew_Error_Count=$cew_Error_Count+1;
                    print("Test Case ERROR (Ncase) in script at line number ", 90, "\n");
                    print("Actual Value is ", $stack3->top(), " \n");
                    print("Expected Value is ", $i, "\n");
                }
             }
             catch{
                $cew_Error_Count=$cew_Error_Count+1;
                my $except = $_;
                if (ref($except) eq "exc::exception") {
                    print("Test Case ERROR (Ncase) in script at line number ", 90, "\n");
                    print("Unexpected Exception full thrown\n");
                }
             }
          };

}

for (my $i=99; $i>=0; $i--) {
   
          do {
            $cew_Test_Count=$cew_Test_Count+1;
            try {
                 ;
                if (($stack3->top()) != ($i)) {
                    $cew_Error_Count=$cew_Error_Count+1;
                    print("Test Case ERROR (Ncase) in script at line number ", 94, "\n");
                    print("Actual Value is ", $stack3->top(), " \n");
                    print("Expected Value is ", $i, "\n");
                }
             }
             catch{
                $cew_Error_Count=$cew_Error_Count+1;
                my $except = $_;
                if (ref($except) eq "exc::exception") {
                    print("Test Case ERROR (Ncase) in script at line number ", 94, "\n");
                    print("Unexpected Exception full thrown\n");
                }
             }
          };

   
          do {
            $cew_Test_Count=$cew_Test_Count+1;
            try {
                $stack3->pop() ;
                if ((0) != (0)) {
                    $cew_Error_Count=$cew_Error_Count+1;
                    print("Test Case ERROR (Ncase) in script at line number ", 95, "\n");
                    print("Actual Value is ", 0, " \n");
                    print("Expected Value is ", 0, "\n");
                }
             }
             catch{
                $cew_Error_Count=$cew_Error_Count+1;
                my $except = $_;
                if (ref($except) eq "exc::exception") {
                    print("Test Case ERROR (Ncase) in script at line number ", 95, "\n");
                    print("Unexpected Exception full thrown\n");
                }
             }
          };

}
   
print("\n**********Summary**********\n");
          print("Total number of test cases = ", $cew_Test_Count, "\n");
          print("Total number of test cases in error = ", $cew_Error_Count, "\n");

File: lab5/IntSetExc1/tb.c
#include "IntSet.h"
#include "Exceptions.h"
#include <stdio.h>













int main() {

	int cew_Test_Count = 0;
	 int cew_Error_Count = 0;
	


	/******************************************************************
	Empty Simple Set 
	******************************************************************/
	TRY {
		 cew_Test_Count = cew_Test_Count + 1;
		 SSInit();SSAdd(10);SSDel(10);SSDel(10);
		 cew_Error_Count = cew_Error_Count + 1;
		 printf("Test Case ERROR (Ecase) in script at line number %d\n", 15);
		 printf("There was no Exception thrown\n");
		 printf("Expected Exception was %s\n", getException(NotFoundExc));
	 }
	 CATCH {
		 if (EXnum__ != (NotFoundExc)) {
			 cew_Error_Count = cew_Error_Count + 1;
			 printf("Test Case ERROR (Ncase) in script at line number %d\n", 15);
			 printf("Unexpected Exception thrown was %s\n", getException(EXnum__));
		 }
	 }
 	 FINALLY
	 ETRY;
	

	TRY {
		 cew_Test_Count = cew_Test_Count + 1;
		 SSInit();SSAdd(10);SSDel(10);
		 if (SSSize() != 0){
			 cew_Error_Count = cew_Error_Count + 1;
			 printf("Test Case ERROR (Ncase) in script at line number %d\n", 16);
			 printf("Actual Value was %d\n", SSSize());
	       printf("Expected Value was %d\n", 0);
		 }
	 }
	 CATCH {
		 cew_Error_Count = cew_Error_Count + 1;
		 printf("Test Case ERROR (Ecase) in script at line number %d\n", 16);
		 printf("Unexpected Exception thrown was %s\n", getException(EXnum__));
	 }
	 FINALLY
	 ETRY;
	

	TRY {
		 cew_Test_Count = cew_Test_Count + 1;
		 SSInit();SSAdd(10);SSDel(10); (1/0);
		 if (SSSize() != 0){
			 cew_Error_Count = cew_Error_Count + 1;
			 printf("Test Case ERROR (Ncase) in script at line number %d\n", 17);
			 printf("Actual Value was %d\n", SSSize());
	       printf("Expected Value was %d\n", 0);
		 }
	 }
	 CATCH {
		 cew_Error_Count = cew_Error_Count + 1;
		 printf("Test Case ERROR (Ecase) in script at line number %d\n", 17);
		 printf("Unexpected Exception thrown was %s\n", getException(EXnum__));
	 }
	 FINALLY
	 ETRY;
	

	TRY {
		 cew_Test_Count = cew_Test_Count + 1;
		 SSInit();SSAdd(10);
		 cew_Error_Count = cew_Error_Count + 1;
		 printf("Test Case ERROR (Ecase) in script at line number %d\n", 18);
		 printf("There was no Exception thrown\n");
		 printf("Expected Exception was %s\n", getException(FullExc));
	 }
	 CATCH {
		 if (EXnum__ != (FullExc)) {
			 cew_Error_Count = cew_Error_Count + 1;
			 printf("Test Case ERROR (Ncase) in script at line number %d\n", 18);
			 printf("Unexpected Exception thrown was %s\n", getException(EXnum__));
		 }
	 }
 	 FINALLY
	 ETRY;
	
 				/* Should Fail*/

	/****************************************************************** 
	Half Full Simple Set 
	******************************************************************/
	TRY {
		 cew_Test_Count = cew_Test_Count + 1;
		 SSInit();SSAdd(9);SSDel(10);SSSize();
		 cew_Error_Count = cew_Error_Count + 1;
		 printf("Test Case ERROR (Ecase) in script at line number %d\n", 23);
		 printf("There was no Exception thrown\n");
		 printf("Expected Exception was %s\n", getException(1));
	 }
	 CATCH {
		 if (EXnum__ != (1)) {
			 cew_Error_Count = cew_Error_Count + 1;
			 printf("Test Case ERROR (Ncase) in script at line number %d\n", 23);
			 printf("Unexpected Exception thrown was %s\n", getException(EXnum__));
		 }
	 }
 	 FINALLY
	 ETRY;
	
	/* Should Fail*/
   TRY {
		 cew_Test_Count = cew_Test_Count + 1;
		 SSInit();SSAdd(10);
		 if (SSSize() != 1){
			 cew_Error_Count = cew_Error_Count + 1;
			 printf("Test Case ERROR (Ncase) in script at line number %d\n", 24);
			 printf("Actual Value was %d\n", SSSize());
	       printf("Expected Value was %d\n", 1);
		 }
	 }
	 CATCH {
		 cew_Error_Count = cew_Error_Count + 1;
		 printf("Test Case ERROR (Ecase) in script at line number %d\n", 24);
		 printf("Unexpected Exception thrown was %s\n", getException(EXnum__));
	 }
	 FINALLY
	 ETRY;
	

	TRY {
		 cew_Test_Count = cew_Test_Count + 1;
		 SSInit();SSAdd(10);
		 if (SSIsMem(10) != 1){
			 cew_Error_Count = cew_Error_Count + 1;
			 printf("Test Case ERROR (Ncase) in script at line number %d\n", 25);
			 printf("Actual Value was %d\n", SSIsMem(10));
	       printf("Expected Value was %d\n", 1);
		 }
	 }
	 CATCH {
		 cew_Error_Count = cew_Error_Count + 1;
		 printf("Test Case ERROR (Ecase) in script at line number %d\n", 25);
		 printf("Unexpected Exception thrown was %s\n", getException(EXnum__));
	 }
	 FINALLY
	 ETRY;
	


	/****************************************************************** 
	Full Simple Set 	
	******************************************************************/
	TRY {
		 cew_Test_Count = cew_Test_Count + 1;
		 SSInit();SSAdd(10);SSAdd(10);
		 cew_Error_Count = cew_Error_Count + 1;
		 printf("Test Case ERROR (Ecase) in script at line number %d\n", 30);
		 printf("There was no Exception thrown\n");
		 printf("Expected Exception was %s\n", getException(DuplicateExc));
	 }
	 CATCH {
		 if (EXnum__ != (DuplicateExc)) {
			 cew_Error_Count = cew_Error_Count + 1;
			 printf("Test Case ERROR (Ncase) in script at line number %d\n", 30);
			 printf("Unexpected Exception thrown was %s\n", getException(EXnum__));
		 }
	 }
 	 FINALLY
	 ETRY;
	

	TRY {
		 cew_Test_Count = cew_Test_Count + 1;
		 SSInit();SSAdd(10);SSAdd(10);
		 cew_Error_Count = cew_Error_Count + 1;
		 printf("Test Case ERROR (Ecase) in script at line number %d\n", 31);
		 printf("There was no Exception thrown\n");
		 printf("Expected Exception was %s\n", getException(FullExc));
	 }
	 CATCH {
		 if (EXnum__ != (FullExc)) {
			 cew_Error_Count = cew_Error_Count + 1;
			 printf("Test Case ERROR (Ncase) in script at line number %d\n", 31);
			 printf("Unexpected Exception thrown was %s\n", getException(EXnum__));
		 }
	 }
 	 FINALLY
	 ETRY;
	
	/* Should Fail*/
	TRY {
		 cew_Test_Count = cew_Test_Count + 1;
		 SSInit();SSAdd(10);SSAdd(11);
		 if (SSSize() != 2){
			 cew_Error_Count = cew_Error_Count + 1;
			 printf("Test Case ERROR (Ncase) in script at line number %d\n", 32);
			 printf("Actual Value was %d\n", SSSize());
	       printf("Expected Value was %d\n", 2);
		 }
	 }
	 CATCH {
		 cew_Error_Count = cew_Error_Count + 1;
		 printf("Test Case ERROR (Ecase) in script at line number %d\n", 32);
		 printf("Unexpected Exception thrown was %s\n", getException(EXnum__));
	 }
	 FINALLY
	 ETRY;
	

	TRY {
		 cew_Test_Count = cew_Test_Count + 1;
		 SSInit();SSAdd(10);SSAdd(11);SSAdd(12);
		 cew_Error_Count = cew_Error_Count + 1;
		 printf("Test Case ERROR (Ecase) in script at line number %d\n", 33);
		 printf("There was no Exception thrown\n");
		 printf("Expected Exception was %s\n", getException(FullExc));
	 }
	 CATCH {
		 if (EXnum__ != (FullExc)) {
			 cew_Error_Count = cew_Error_Count + 1;
			 printf("Test Case ERROR (Ncase) in script at line number %d\n", 33);
			 printf("Unexpected Exception thrown was %s\n", getException(EXnum__));
		 }
	 }
 	 FINALLY
	 ETRY;
	


	printf("\n**********Summary**********\n");
	 printf("Total number of test cases = %d\n", cew_Test_Count);
	 printf("Total number of test cases in error = %d\n", cew_Error_Count);
	

	
	return 0;
}
